helper/cert/source_gen.go

```diff
@@ -182,30 +184,38 @@ func (s *GenSource) checkLeader(ctx context.Context, changed chan<- bool) {
 	}
 }
 
-func (s *GenSource) retryUpdateSecret(ctx context.Context, leaderCh chan bool, bundle Bundle) {
-	// New exponential backoff
+func (s *GenSource) retryUpdateSecret(ctx context.Context, leaderCh chan bool, bundle Bundle) error {
+	// New exponential backoff, with a max elapsed time of 90% of the newly
+	// created cert expiration, since that's the point where it would be renewed
+	// by the calling function
 	bo := backoff.NewExponentialBackOff()
 	bo.MaxInterval = 30 * time.Second
-	bo.MaxElapsedTime = 30 * time.Minute
+	bo.MaxElapsedTime = time.Duration(float64(s.expiry()) * 0.90)
 	ticker := backoff.NewTicker(bo)
 	defer ticker.Stop()
-	for {
+
+	for range ticker.C {
```

Comment:
I'm not sure how important this is, but I think this `range` loop means we may wait up to `bo.MaxInterval` before returning when a leadership change happens or the context is cancelled. I think you can fix this by switching back to the same `for`/`select` pattern as before with a small modification to the `ticker.C` case:

```go
var err error
for {
	select {
	case _, ok := <-ticker.C:
		if !ok {
			// channel closed implies timeout
			return fmt.Errorf("timed out attempting to update cert secret: %w", err)
		}
		err = s.updateSecret(...)
	case ...:
		...
	// No default case
}
```

Reply:
Ah yes! I forgot you could do that with channels. Changed in 8cded14.